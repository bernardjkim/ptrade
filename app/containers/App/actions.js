/*
 *
 * App actions
 *
 */

import {
  LOAD_TOKEN,
  CREATE_SESSION,
  CREATE_SESSION_ERROR,
  CREATE_SESSION_SUCCESS,
  DELETE_SESSION,
} from './constants';

/**
 * Load current jwt token if available
 *
 * @return {object} An action object with a type of LOAD_TOKEN
 *
 */
export function loadToken() {
  const token = sessionStorage.getItem('jwtToken');
  return {
    type: LOAD_TOKEN,
    token,
  };
}

/**
 * Delete the current session token
 *
 * @return {object} An action object with a type of DELETE_TOKEN
 *
 */
export function deleteSession() {
  sessionStorage.removeItem('jwtToken');
  return {
    type: DELETE_SESSION,
  };
}

/**
 * Create session, this action starts the request saga
 *
 * @return {object} An action object with a type of CREATE_SESSION
 */
export function createSession() {
  return {
    type: CREATE_SESSION,
  };
}

/**
 * Dispatched when the session token is loaded by the request saga
 *
 * @param  {string} token The chart data
 *
 * @return {object}       An action object with a type of CREATE_SESSION_SUCCESS
 *                        passing the token
 */
export function createSessionSuccess(token) {
  return {
    type: CREATE_SESSION_SUCCESS,
    token,
  };
}

/**
 * Dispatched when createing the session fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of CREATE_SESSION_ERROR
 *                        passing the error
 */
export function createSessionError(error) {
  return {
    type: CREATE_SESSION_ERROR,
    error,
  };
}
