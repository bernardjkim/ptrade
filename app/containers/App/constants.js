/*
 * AppConstants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */

// CREATE SESSION
export const CREATE_SESSION = 'app/App/CREATE_SESSION';
export const CREATE_SESSION_SUCCESS = 'app/App/CREATE_SESSION_SUCCESS';
export const CREATE_SESSION_ERROR = 'app/App/CREATE_SESSION_ERROR';

// DELETE SESSION
// TODO: is it enough to just delete the client-side token? or do i need to
// black list token on the backend?
export const DELETE_SESSION = 'app/App/DELETE_SESSION';
export const DELETE_SESSION_SUCCESS = 'app/App/DELETE_SESSION_SUCCESS';
export const DELETE_SESSION_ERROR = 'app/App/DELETE_SESSION_ERROR';

export const LOAD_TOKEN = 'app/App/LOAD_TOKEN';
