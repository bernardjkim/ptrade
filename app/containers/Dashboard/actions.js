/*
 *
 * Dashboard actions
 *
 */

import {
  CHANGE_SEARCH,
  CHANGE_TIME_FRAME,
  SELECT_SYMBOL,
  LOAD_CHART,
  LOAD_CHART_SUCCESS,
  LOAD_CHART_ERROR,
  LOAD_QUOTE,
  LOAD_QUOTE_SUCCESS,
  LOAD_QUOTE_ERROR,
} from './constants';

/**
 * Change the current search symbol
 *
 * @param  {string} search  A stock symbol
 *
 * @return {object}         An action object with a type of CHANGE_SEARCH
 */
export function changeSearch(search) {
  return {
    type: CHANGE_SEARCH,
    search,
  };
}

/**
 * Change the current selected symbol
 *
 * @param  {string} symbol  Stock symbol
 *
 * @return {object}         An action object with a type of SELECT_SYMBOL
 */
export function selectSymbol(symbol) {
  return {
    type: SELECT_SYMBOL,
    symbol,
  };
}

/**
 * Change the selected time frame
 *
 * @param  {number} tf  Time frame index
 *
 * @return {object}     An action object with a type of CHANGE_TIME_FRAME
 */
export function changeTimeFrame(tf) {
  return {
    type: CHANGE_TIME_FRAME,
    tf,
  };
}

/**
 * Load chart, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_CHART
 */
export function loadChart() {
  return {
    type: LOAD_CHART,
  };
}

/**
 * Dispatched when the chart is loaded by the request saga
 *
 * @param  {array} chart  The chart data
 *
 * @return {object}       An action object with a type of LOAD_CHART_SUCCESS passing the data
 */
export function chartLoaded(chart) {
  return {
    type: LOAD_CHART_SUCCESS,
    chart,
  };
}

/**
 * Dispatched when loading the chart fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_CHART_ERROR passing the error
 */
export function chartLoadingError(error) {
  return {
    type: LOAD_CHART_ERROR,
    error,
  };
}

/**
 * Load the stock quote, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_QUOTE
 */
export function loadQuote() {
  return {
    type: LOAD_QUOTE,
  };
}

/**
 * Dispatched when the stock quote is loaded by the request saga
 *
 * @param  {object} quote The stock quote
 *
 * @return {object}       An action object with a type of LOAD_QUOTE_SUCCESS passing the quote
 */
export function quoteLoaded(quote) {
  return {
    type: LOAD_QUOTE_SUCCESS,
    quote,
  };
}

/**
 * Dispatched when loading the stock quote fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_QUOTE_ERROR passing the error
 */
export function quoteLoadingError(error) {
  return {
    type: LOAD_QUOTE_ERROR,
    error,
  };
}
