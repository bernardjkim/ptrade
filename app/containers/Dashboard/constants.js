/*
 *
 * Dashboard constants
 *
 */

export const CHANGE_SEARCH = 'app/Dashboard/CHANGE_SEARCH';
export const CHANGE_TIME_FRAME = 'app/Dashboard/CHANGE_TIME_FRAME';

export const LOAD_STOCK_DATA = 'app/Dashboard/LOAD_STOCK_DATA';
export const LOAD_STOCK_DATA_SUCCESS = 'app/Dashboard/LOAD_STOCK_DATA_SUCCESS';
export const LOAD_STOCK_DATA_ERROR = 'app/Dashboard/LOAD_STOCK_DATA_ERROR';
