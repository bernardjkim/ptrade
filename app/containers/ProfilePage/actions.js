/*
 *
 * ProfilePage actions
 *
 */

import {
  CHANGE_TIME_FRAME,
  CHANGE_TRANSFER_AMOUNT,
  LOAD_BALANCE,
  LOAD_BALANCE_ERROR,
  LOAD_BALANCE_SUCCESS,
  LOAD_POSITIONS,
  LOAD_POSITIONS_ERROR,
  LOAD_POSITIONS_SUCCESS,
  LOAD_CHART,
  LOAD_CHART_ERROR,
  LOAD_CHART_SUCCESS,
  REQUEST_TRANSFER,
  REQUEST_TRANSFER_ERROR,
  REQUEST_TRANSFER_SUCCESS,
} from './constants';

export function changeTimeFrame(timeFrame) {
  return {
    type: CHANGE_TIME_FRAME,
    timeFrame,
  };
}

export function changeTransferAmount(amount) {
  return {
    type: CHANGE_TRANSFER_AMOUNT,
    amount,
  };
}

export function loadBalance() {
  return {
    type: LOAD_BALANCE,
  };
}

export function loadBalanceSuccess(balance) {
  return {
    type: LOAD_BALANCE_SUCCESS,
    balance,
  };
}

export function loadBalanceError(error) {
  return {
    type: LOAD_BALANCE_ERROR,
    error,
  };
}

export function loadPositions() {
  return {
    type: LOAD_POSITIONS,
  };
}

export function loadPositionsSuccess(positions) {
  let total = 0;
  positions.forEach(p => {
    total += p.shares * p.price_per_share;
  });
  return {
    type: LOAD_POSITIONS_SUCCESS,
    positions,
    total,
  };
}

export function loadPositionsError(error) {
  return {
    type: LOAD_POSITIONS_ERROR,
    error,
  };
}

export function loadChart() {
  return {
    type: LOAD_CHART,
  };
}

export function loadChartSuccess(chart) {
  return {
    type: LOAD_CHART_SUCCESS,
    chart,
  };
}

export function loadChartError(error) {
  return {
    type: LOAD_CHART_ERROR,
    error,
  };
}

export function requestTransfer() {
  return {
    type: REQUEST_TRANSFER,
  };
}

export function requestTransferSuccess() {
  return {
    type: REQUEST_TRANSFER_SUCCESS,
  };
}

export function requestTransferError(error) {
  return {
    type: REQUEST_TRANSFER_ERROR,
    error,
  };
}
