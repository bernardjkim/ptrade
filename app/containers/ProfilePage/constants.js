/*
 *
 * ProfilePage constants
 *
 */

export const CHANGE_TIME_FRAME = 'app/ProfilePage/CHANGE_TIME_FRAME';
export const SET_PORTFOLIO_VALUE = 'app/ProfilePage/SET_PORTFOLIO_VALUE';

export const LOAD_BALANCE = 'app/ProfilePage/LOAD_BALANCE';
export const LOAD_BALANCE_SUCCESS = 'app/ProfilePage/LOAD_BALANCE_SUCCCESS';
export const LOAD_BALANCE_ERROR = 'app/ProfilePage/LOAD_BALANCE_ERROR';

export const LOAD_POSITIONS = 'app/ProfilePage/LOAD_POSITIONS';
export const LOAD_POSITIONS_SUCCESS = 'app/ProfilePage/LOAD_POSITIONS_SUCCESS';
export const LOAD_POSITIONS_ERROR = 'app/ProfilePage/LOAD_POSITIONS_ERROR';

export const LOAD_CHART = 'app/ProfilePage/LOAD_CHART';
export const LOAD_CHART_SUCCESS = 'app/ProfilePage/LOAD_CHART_SUCCESS';
export const LOAD_CHART_ERROR = 'app/ProfilePage/LOAD_CHART_ERROR';
