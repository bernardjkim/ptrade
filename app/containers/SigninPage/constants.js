/*
 *
 * SigninPage constants
 *
 */

export const CHANGE_INPUT = 'app/SigninPage/CHANGE_INPUT';
