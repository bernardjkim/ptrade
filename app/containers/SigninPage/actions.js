/*
 *
 * SigninPage actions
 *
 */

import { CHANGE_INPUT } from './constants';

/**
 * Change the input value for specified field
 *
 * @param  {string} field The field to be changed
 * @param  {string} value The value to be set
 *
 * @return {object}       An object of type CHANGE_INPUT
 */
export function changeInput(field, value) {
  return {
    type: CHANGE_INPUT,
    field,
    value,
  };
}
