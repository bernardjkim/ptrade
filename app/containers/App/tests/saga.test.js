/**
 * Test sagas
 */

/* eslint-disable redux-saga/yield-effects */
import { takeLatest, put } from 'redux-saga/effects';
import { CREATE_SESSION } from '../constants';
import { createSessionSuccess, createSessionError } from '../actions';
import appSaga, { getSession } from '../saga';

const email = 'test@email.com';
const password = 'password';

describe('createSession Saga', () => {
  let createSessionGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    createSessionGenerator = getSession();

    const selectDescriptor = createSessionGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    createSessionGenerator.next(email);
    const callDescriptor = createSessionGenerator.next(password).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('request the token successfully', () => {
    const token = 'token';
    const putDescriptor = createSessionGenerator.next(token).value;
    expect(putDescriptor).toEqual(put(createSessionSuccess(token)));
  });

  it('should call the createSessionError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = createSessionGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(createSessionError(response)));
  });
});

describe('Saga', () => {
  it('should start task to watch for global actions', () => {
    const takeLatestDescriptor = appSaga().next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(CREATE_SESSION, getSession),
    );
  });
});
