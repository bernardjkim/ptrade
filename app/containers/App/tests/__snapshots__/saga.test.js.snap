// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSession Saga request the token successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`createSession Saga request the token successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://ptrade-api-staging.herokuapp.com/v1/sessions",
      Object {
        "body": "email=test%40email.com&password=password",
        "headers": Object {
          "Content-Type": "application/x-www-form-urlencoded",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`createSession Saga should call the createSessionError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`createSession Saga should call the createSessionError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://ptrade-api-staging.herokuapp.com/v1/sessions",
      Object {
        "body": "email=test%40email.com&password=password",
        "headers": Object {
          "Content-Type": "application/x-www-form-urlencoded",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
